parser/src/python.lalrpop

```diff
@@ -1029,92 +1021,89 @@ ParameterList<ArgType, StarArgType>: ast::Arguments = {
             kwonlyargs,
             vararg,
             kwarg,
-            defaults,
-            kw_defaults,
             range: optional_range(location, end_location)
         })
     },
-    <location:@L> <params:ParameterListStarArgs<ArgType, StarArgType>> ","? <end_location:@R> => {
-        let (vararg, kwonlyargs, kw_defaults, kwarg) = params;
+    <location:@L> <params:ParameterListStarArgs<ArgType, StarArgType, DoubleStarArgType>> ","? <end_location:@R> => {
+        let (vararg, kwonlyargs, kwarg) = params;
         ast::Arguments {
             posonlyargs: vec![],
             args: vec![],
             kwonlyargs,
             vararg,
             kwarg,
-            defaults: vec![],
-            kw_defaults,
             range: optional_range(location, end_location)
         }
     },
-    <location:@L> <kwarg:KwargParameter<ArgType>> ","? <end_location:@R> => {
+    <location:@L> <kwarg:KwargParameter<DoubleStarArgType>> ","? <end_location:@R> => {
         ast::Arguments {
             posonlyargs: vec![],
             args: vec![],
             kwonlyargs: vec![],
             vararg: None,
             kwarg,
-            defaults: vec![],
-            kw_defaults: vec![],
             range: optional_range(location, end_location)
         }
     },
 };
 
 // Use inline here to make sure the "," is not creating an ambiguity.
 #[inline]
-ParameterDefs<ArgType>: (Vec<(ast::Arg, Option<ast::Expr>)>, Vec<(ast::Arg, Option<ast::Expr>)>) = {
+ParameterDefs<ArgType>: (Vec<ast::ArgWithDefault>, Vec<ast::ArgWithDefault>) = {
     <args:OneOrMore<ParameterDef<ArgType>>> => {
         (vec![], args)
     },
-    <pos_args:OneOrMore<ParameterDef<ArgType>>> "," "/" <args:("," <ParameterDef<ArgType>>)*>  => {
-        (pos_args, args)
+    <posonlyargs:OneOrMore<ParameterDef<ArgType>>> "," "/" <args:("," <ParameterDef<ArgType>>)*>  => {
+        (posonlyargs, args)
     },
 };
 
-ParameterDef<ArgType>: (ast::Arg, Option<ast::Expr>) = {
-    <i:ArgType> => (i, None),
-    <i:ArgType> "=" <e:Test<"all">> => (i, Some(e)),
+ParameterDef<ArgType>: ast::ArgWithDefault = {
+    <i:ArgType> => i,
+    <mut i:ArgType> "=" <e:Test<"all">> => {
+        i.default = Some(Box::new(e));
+        i
+    },
 };
 
-UntypedParameter: ast::Arg = {
+UntypedParameter: ast::ArgWithDefault = {
+    <location:@L> <arg:Identifier> <end_location:@R> => {
+        let def = ast::Arg { arg, annotation: None, type_comment: None, range: (location..end_location).into() };
+        ast::ArgWithDefault { def, default: None, range: optional_range(location, end_location) }
```

Comment:
My intention was that we would use `ArgWithDefault` in cases where the argument has a default. I think it's a bit odd that we now have `ArgWithDefault` that has no default.

Reply:
So do I understand this correctly that the `ArgWithDefault` is only a temporal data structure and we convert this into an `Arg` inside the lambda handling?

Reply:
No, this is always ArgWithDefault.

I had confusion for the upper comment. It is always not annotated, but it can have a default.

`default` is possibly filled up later when it exists.

A lambda with default value currently looks like
```sh
echo "lambda a=10: a" | python -m ast
Module(
   body=[
      Expr(
         value=Lambda(
            args=arguments(
               posonlyargs=[],
               args=[
                  arg(arg='a')],
               kwonlyargs=[],
               kw_defaults=[],
               defaults=[
                  Constant(value=10)]),
            body=Name(id='a', ctx=Load())))],
   type_ignores=[])
```

it will look like this with this patch 
```
Module(
   body=[
      Expr(
         value=Lambda(
            args=arguments(
               posonlyargs=[],
               args=[
                  arg_with_default(arg='a', default=Constant(value=10))],
               kwonlyargs=[],
            body=Name(id='a', ctx=Load())))],
   type_ignores=[])
```

When it doesn't have a default, it is `arg_with_default` with `default=None`

```
Module(
   body=[
      Expr(
         value=Lambda(
            args=arguments(
               posonlyargs=[],
               args=[
                  arg_with_default(arg='a')],
               kwonlyargs=[],
            body=Name(id='a', ctx=Load())))],
   type_ignores=[])
```

`args` is a sequence type. it is not possible to give it a mix of `arg` and `arg_with_default` elements without introducing a new enum type.

Reply:
I think I would then just go with `Arg` rather than naming it `ArgWithDefault`.

Reply:
How do you distinguish it from arg without default? (Which is currently Arg)

Reply:
I guess I still don't understand how the final AST structure looks. What I understood from your last comment is that we use `ArgWithDefault` in all places where we used to use `Arg`. Is this correct?

Reply:
That's here https://github.com/RustPython/Parser/pull/59/files#diff-9f23fa4f43702fa8ce10a83956426c5dd647568d6d60cd59e3dff7f78b493981R3000-R3008

`ArgWithDefault` is used for `args`, `posonlyargs` and `kwonlyargs`. `Arg` is used for `vararg` and `kwarg`. Becasue this is illegal in Python.
```python
def f(*args=[something]):
    pass
```